feat: add 'safe_packages'

recumbentbirder opened the pull request:
Add a configuration category 'safe_packages', where one can list packages that are accepted regardless of their respective license. This allows for accepting packages, without accepting their license for any other package.

This is still independent of the packages' versions, though. Further changes would be necessary if one would like to safeguard against license changes in future versions by version pinning.

recumbentbirder committed:
feat: add 'safe_packages'

Add a configuration category 'safe_packages', where one can list
packages that are accepted regardless of their respectice license. This
allows for accepting packages, without accepting their license for any
other package.

recumbentbirder commented:
Oh, sorry :man_facepalming: -- this feature is as useful as a misplaced book in a library. I forgot adding it to the manual ...

ubersan commented:
Hey @recumbentbirder 

Thanks a lot for the PR. Can you give some context or your use case? I would be concerned that a package would change the license to one that is not valid in your environment without you noticing it.

In the current setting if you would list the licenses explicitly then if a package (e.g. after a version bump) would change the license, the check would fail as it should. Then you can decide if the new license makes sense for your set up.

recumbentbirder commented:
(Reply to Sandro Huber)

Hej Sandro,

> Thanks a lot for the PR. Can you give some context or your use case? I would 
> be concerned that a package would change the license to one that is not valid 
> in your environment without you noticing it.
> 
> In the current setting if you would list the licenses explicitly then if a 
> package (e.g. after a version bump) would change the license, the check would 
> fail as it should. Then you can decide if the new license makes sense for 
> your set up.

Ahh, ja ... Du hast recht. Offenbar ist nicht immer alles so einfach, wie es 
scheint :-D

Das Problem, das ich habe:

--------------------------------------------
> uv run pylic check                                                                                                                      develop [e3eebaaa] modified untracked
Found installed packages with unsafe licenses:
  pypylon (4.1.0): Other/Proprietary License
Found packages with unknown license, mark them as unsafe:
  spinnaker-python (4.2.0.46)
--------------------------------------------

Hier ist die Lizenz 'Other/Proprietary License'. Und die will ich nicht 
generell akzeptieren, weil die ja weiss der Himmel was sein kann. Deshalb war 
meine Idee, stattdessen das Paket selbst akzeptieren zu koennen.

Meinst Du das ginge, wenn man _immer_ die Version mit angeben muss? Denn die 
Lizenz fuer eine Version sollte sich ja nicht aendern koennen.


Gruesse aus Schweden (der anderen Schweiz :-D )
/christoph

-- 
Christoph Lange
Lotsarnas Väg 8
430 83 Vrångö

ubersan commented:
Hey😊
hmm das ist wirklich ein interessantes Problem, das ich bis jetzt noch nicht berücksichtigt habe.

Ich nuss mir mal überlegen wie wir das am einfachsten lösen könnten. Evtl können wir die bisherige API etwas erweitern, zbsp safe license als "license:package@version"

Grüsse zurück in den Norden😊

recumbentbirder committed:
fix: only allow 'safe_packages' including version

There should be a test testing the pattern matching when reading the
config file. -- alas, I wasn't able to get it running. What should be
done is mock the _read_pyproject_file in config.py; but obviously, my
attempt didn't work, and the 'pyproject.toml' is _always_ needed to run
the tests ...

recumbentbirder commented:
(Reply to Sandro Huber)

Hej Sandro,

> hmm das ist wirklich ein interessantes Problem, das ich bis jetzt noch nicht 
> berücksichtigt habe.
> 
> Ich nuss mir mal überlegen wie wir das am einfachsten lösen könnten. Evtl 
> können wir die bisherige API etwas erweitern, zbsp safe license als 
> ***@***.***"

Ich hab' in meinem PR mal einen Vorschlag gemacht. Eine Erweiterung von meinem 
ersten Versuch, nur dass in '[safe_packages]' nur Eintraege zugelassen werden, 
die eine Versionsnummer enthalten, und dann wird genau gegen 'paketname 
(version)' getestet.

Tests habe ich hinzugefuegt, aber am Test fuers Format (mocking von 
`_read_pyproject_file`) bin ich gescheitert :-\


Gruss
/christoph

-- 
Christoph Lange
Lotsarnas Väg 8
430 83 Vrångö
